src/ir_passes/phantom_check.rs

```diff
@@ -0,0 +1,127 @@
+use itertools::Itertools;
+
+use crate::{
+    diagnostics,
+    errors::Error,
+    ir::{self, Ctx},
+    ir_visitor::{Action, Visitor, VisitorData},
+    utils::GPosIdx,
+};
+use std::collections::HashSet;
+
+#[derive(Default)]
+/// Checks if a user-level phantom events are valid.
+/// Phantom events are valid iff:
+/// 1. The component doesn't share any instances
+/// 2. The component doesn't use an subcomponents that need to use the
+///    corresponding event in their interface, i.e., the uses of the event are all phantom
+pub struct PhantomCheck {
+    instance_used: HashSet<ir::InstIdx>,
+    prev_invoke: Option<ir::InvIdx>,
+    phantom_events: Vec<ir::EventIdx>,
+    diag: diagnostics::Diagnostics,
+    diag_count: u32,
+}
+
+impl Visitor for PhantomCheck {
+    fn name() -> &'static str {
+        "phantom-check"
+    }
+
+    fn start(&mut self, data: &mut VisitorData) -> Action {
+        let comp = &data.comp;
+        self.phantom_events = comp.phantom_events().collect();
+        if self.phantom_events.is_empty() {
+            Action::Stop
+        } else {
+            Action::Continue
+        }
+    }
+
+    fn invoke(&mut self, inv: ir::InvIdx, data: &mut VisitorData) -> Action {
+        // Check if the instance has already been used
+        let comp = &data.comp;
+        let ctx = data.ctx();
+        let inst = comp.get(inv).inst;
+        if self.instance_used.get(&inst).is_some() {
+            for (time, info) in inv.times(comp) {
+                if let Some(e) =
+                    self.phantom_events.iter().find(|e| time.event(comp) == **e)
+                {
+                    let bind_info =
+                        comp.get(info).as_event_bind().unwrap().bind_loc;
+                    let event_info = comp
+                        .get(comp.get(*e).info)
+                        .as_event()
+                        .unwrap()
+                        .bind_loc;
+                    let prev_inv_info = comp
+                        .get(comp.get(self.prev_invoke.unwrap()).info)
+                        .as_invoke()
+                        .unwrap()
+                        .inst_loc;
+                    let err =
+                        Error::malformed("reuses instance uses phantom event for scheduling")
+                    .add_note(self.diag.add_info("invocation uses phantom event", bind_info))
+                    .add_note(self.diag.add_info("event is a phantom event", event_info))
+                    .add_note(self.diag.add_info("previous use", prev_inv_info))
+                    .add_note(self.diag.add_info("phantom ports are compiled away and cannot be used for resource sharing", GPosIdx::UNKNOWN));
+                    self.diag.add_error(err);
+                    self.diag_count += 1;
+                    return Action::Stop;
```

Comment:
If we return `Stop` here, we won't be able to report other phantom check error message in this component right?